[SPARK-34484][SQL] Rename `map` to `mapAttr` in Catalyst DSL

sarutak opened the pull request:
### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This PR proposes to rename `map` to `mapAttr` in `dsl/package.scala` .
This PR also removes APIs which take `Symbol` type parameters.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
With the Catalyst DSL (`dsl/package.scala`), we have two ways to represent attributes.

1. Symbol literals (`'` syntax)
2. `$""` syntax which is defined in `sql/catalyst` module using string context.

But they have problems.

Regarding symbol literals, the scala community deprecates the symbol literals in Scala 2.13. We could alternatively use `Symbol` constructor but what is worse, Scala will completely remove `Symbol` in the future (https://scalacenter.github.io/scala-3-migration-guide/docs/incompatibilities/dropped-features.html).


> Although scala.Symbol is useful for migration, beware that it is deprecated and that it will be removed from the scala-library. You are recommended, as a second step, to replace them with plain string literals "xwy" or a dedicated class.

Regarding `$""` syntax, this has two problems.
The first problem is that the syntax conflicts with another `$""` syntax defined in `sql/core` module.
You can easily see the problem with the Spark Shell.

```
import org.apache.spark.sql.catalyst.dsl.expressions._
val attr1 = $"attr1"

       error: type mismatch;
        found   : StringContext
        required: ?{def $: ?}
       Note that implicit conversions are not applicable because they are ambiguous:
        both method StringToColumn in class SQLImplicits of type (sc: StringContext): spark.implicits.StringToColumn
        and method StringToAttributeConversionHelper in trait ExpressionConversions of type (sc: StringContext): org.apache.spark.sql.catalyst.dsl.expressions.StringToAttributeConversionHelper
        are possible conversion functions from StringContext to ?{def $: ?}

```

The second problem is that we can't write like `$"attr".map(StringType, StringType)`, though we can write `'attr.map(StringType, StringType)`.
This seems to be a bug of the Scala compiler and will be fixed in neither `2.12` nor `2.13` (https://github.com/scala/scala/pull/7396).

Actually, I'm working on replacing all the symbol literals with `$""` syntax in SPARK-34443 (#31569) and I found this problem in the following test code.

* EncoderResolutionSuite.scala
* ComplexTypeSuite.scala
* ObjectExpressionsSuite.scala
* NestedColumnAliasingSuite.scala
* ReplaceNullWithFalseInPredicateSuite.scala
* SimplifyCastsSuite.scala
* SimplifyConditionalSuite.scala

```
[error] /home/kou/work/oss/spark-scala-2.13/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/EncoderResolutionSuite.scala:212:28: too many arguments (found 2, expected 1) for method map: (f: org.apache.spark.sql.catalyst.expressions.Expression => A): Seq[A]
[error]       $"a".map(StringType, StringType)).foreach { attr =>
```

As a solution for those problems, I propose to rename `map` to `mapAttr`.
With this solution, we can write `"name".attr.mapAttr` rather than `$"name".map` or `'name.map`.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No. `dsl/package.scala` is for internal use.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Modified the tests.

sarutak committed:
Introduce DslAttributeSymbol.

sarutak committed:
Modify tests to use attr() rather than ' syntax.

SparkQA commented:
**[Test build #135315 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135315/testReport)** for PR 31601 at commit [`5b9c205`](https://github.com/apache/spark/commit/5b9c205854ef0082703f2cc7273e5c407926a0f4).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39895/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39895/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39895/

SparkQA commented:
**[Test build #135315 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135315/testReport)** for PR 31601 at commit [`5b9c205`](https://github.com/apache/spark/commit/5b9c205854ef0082703f2cc7273e5c407926a0f4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135315/

sarutak commented:
cc: @cloud-fan

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/EncoderResolutionSuite.scala:
```diff
@@ -54,38 +54,38 @@ class EncoderResolutionSuite extends PlanTest {
 
     // int type can be up cast to long type
-    val attrs1 = Seq('a.string, 'b.int)
+    val attrs1 = Seq(attr("a").string, attr("b").int)
```
We already have `"name".attr` (See the implicit `DslString`), can we extend it to support `"name".attr.string`?

sarutak replied:
Actually, tt's not enough. `"name".attr` returns `UnresolvedAttribute` so we still fail to `"name".attr.map`.

cloud-fan replied:
My general idea is, previous the DSL is based on symbol, can we update the DSL (those implicts) so that it's based on string? Ideally we should simply replace `'abc` to `"abc"` in the test code.

sarutak replied:
It was my first idea but I thought it might be confusable that `"abc"` with implicit conversion doesn't look like an `attribute`. But DSL is for internal so it might be acceptable. What do you think?

sarutak replied:
>  Ideally we should simply replace 'abc to "abc" in the test code.

Yes ideally it should be like what you suggested but I found it's difficult.
`DslString` implements `ImplicitOperators` but `DslSymbol` implements `ImplicitAttribute`. The implementation of `expr` conflicts. `DslString` implements `expr` as `Literal` but `DslSymbol` implements it as `UnresolvedExpression`.

So, the next option would be introduce a new syntax like `attr""` with `StringContext`. 

I said in the description like as follows.
> Another solution would be to introduce a new string interpolation syntax but it's difficult to have workaround when the syntax is conflict with another string interpolation syntax like that $"" defined in catalyst and core conflict so I don't adopt this solution.

But if this is internal usage only, it might be acceptable.

sarutak committed:
Use attr"" rather than attr()

SparkQA commented:
**[Test build #135348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135348/testReport)** for PR 31601 at commit [`c87915d`](https://github.com/apache/spark/commit/c87915d7f6908e89a1ce35444706c06313904f24).

SparkQA commented:
**[Test build #135348 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135348/testReport)** for PR 31601 at commit [`c87915d`](https://github.com/apache/spark/commit/c87915d7f6908e89a1ce35444706c06313904f24).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

sarutak committed:
Fix test.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135348/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39929/

SparkQA commented:
**[Test build #135351 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135351/testReport)** for PR 31601 at commit [`9f78964`](https://github.com/apache/spark/commit/9f78964a52b49323a443495e72bc13c559cd865a).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39929/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39929/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39931/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39931/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39931/

sarutak commented:
cc: @dongjoon-hyun and @HyukjinKwon too for other people's opinion.

SparkQA commented:
**[Test build #135351 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135351/testReport)** for PR 31601 at commit [`9f78964`](https://github.com/apache/spark/commit/9f78964a52b49323a443495e72bc13c559cd865a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135351/

cloud-fan commented:
instead of creating a new syntax, can we use `"abc".attr.int` to do the same thing?

cloud-fan commented:
I agree that the `$"abc"` is not great as it conflicts with the well-known syntax defined in sql/core. But I do like the existing `"abc".attr` syntax instead of the new one. Can we rename `map` to `mapAttr` to avoid the scala bug?

sarutak committed:
Merge branch 'master' of https://github.com/apache/spark into dsl-attr

sarutak committed:
Rename `map` to `mapAttr` in dsl/package.scala.

sarutak commented:
Thanks for the advice. Renamed `map` to `mapAttr`. It seems become not symmetric with `struct` and `array` but it can be acceptable as it's for internal usage.
I also updated the description.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39992/

SparkQA commented:
**[Test build #135412 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135412/testReport)** for PR 31601 at commit [`b306786`](https://github.com/apache/spark/commit/b3067863c0f62c13d869683fbb22464ff2f7ce97).

SparkQA commented:
**[Test build #135412 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135412/testReport)** for PR 31601 at commit [`b306786`](https://github.com/apache/spark/commit/b3067863c0f62c13d869683fbb22464ff2f7ce97).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135412/

sarutak committed:
Fix style.

SparkQA commented:
**[Test build #135421 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135421/testReport)** for PR 31601 at commit [`fb6041e`](https://github.com/apache/spark/commit/fb6041e4bb2284c5dbb0f5f708998b9ba42b603b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40001/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40001/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40001/

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala:
```diff
@@ -308,10 +308,10 @@ package object dsl {
-      def map(mapType: MapType): AttributeReference =
+      def mapAttr(mapType: MapType): AttributeReference =
         AttributeReference(s, mapType, nullable = true)()
 
```
If you have fixed all the places, shall we remove DSL APIs that take Symbol?

sarutak replied:
I've removed and I found most symbol literals are used with DSL APIs (changed 114 files)...

cloud-fan reviewed (approved):
LGTM, please update the PR title

HyukjinKwon reviewed (approved):
I like this approach too. Looks good

SparkQA commented:
**[Test build #135421 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135421/testReport)** for PR 31601 at commit [`fb6041e`](https://github.com/apache/spark/commit/fb6041e4bb2284c5dbb0f5f708998b9ba42b603b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135421/

sarutak committed:
Remove APIs which take symbols from dsl/packages.scala.

SparkQA commented:
**[Test build #135437 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135437/testReport)** for PR 31601 at commit [`71bf596`](https://github.com/apache/spark/commit/71bf59665af3e2b2daffbddd5c8d7621575b859b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40017/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40017/

sarutak committed:
Fix tests.

SparkQA commented:
**[Test build #135440 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135440/testReport)** for PR 31601 at commit [`1c689ae`](https://github.com/apache/spark/commit/1c689ae7dd90506e66c71014bc8df4301431321e).

SparkQA commented:
**[Test build #135437 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135437/testReport)** for PR 31601 at commit [`71bf596`](https://github.com/apache/spark/commit/71bf59665af3e2b2daffbddd5c8d7621575b859b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135437/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40020/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40020/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40020/

SparkQA commented:
**[Test build #135440 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135440/testReport)** for PR 31601 at commit [`1c689ae`](https://github.com/apache/spark/commit/1c689ae7dd90506e66c71014bc8df4301431321e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135440/

sarutak committed:
Further fix tests.

SparkQA commented:
**[Test build #135448 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135448/testReport)** for PR 31601 at commit [`a7866ce`](https://github.com/apache/spark/commit/a7866ce92bb056b4c86d4503a04cdff535ef0068).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40028/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40028/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40028/

sarutak committed:
Further fix tests.

SparkQA commented:
**[Test build #135450 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135450/testReport)** for PR 31601 at commit [`a9e3490`](https://github.com/apache/spark/commit/a9e34901612042997441a767f2501df2cd7b3729).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40030/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40030/

SparkQA commented:
**[Test build #135448 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135448/testReport)** for PR 31601 at commit [`a7866ce`](https://github.com/apache/spark/commit/a7866ce92bb056b4c86d4503a04cdff535ef0068).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135450/

cloud-fan reviewed (approved):
LGTM. This is a lot of work, thanks!

sarutak commented:
retest this please.

cloud-fan commented:
retest this please

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40071/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40071/

SparkQA commented:
**[Test build #135490 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135490/testReport)** for PR 31601 at commit [`a9e3490`](https://github.com/apache/spark/commit/a9e34901612042997441a767f2501df2cd7b3729).

SparkQA commented:
**[Test build #135490 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135490/testReport)** for PR 31601 at commit [`a9e3490`](https://github.com/apache/spark/commit/a9e34901612042997441a767f2501df2cd7b3729).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135490/

sarutak committed:
Merge branch 'master' of https://github.com/apache/spark into dsl-attr

sarutak commented:
Newly added `RemoveNoopUnionSuite` seems to use symbol literal, which causes the build failure.
I'll fix it too.

sarutak committed:
Replace symbol literals in RemoveNoopUnionSuite.

SparkQA commented:
**[Test build #135492 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135492/testReport)** for PR 31601 at commit [`e3dead4`](https://github.com/apache/spark/commit/e3dead46d8e49d137aab17a6c793f9913c4c5405).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40073/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40073/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40073/

SparkQA commented:
**[Test build #135492 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135492/testReport)** for PR 31601 at commit [`e3dead4`](https://github.com/apache/spark/commit/e3dead46d8e49d137aab17a6c793f9913c4c5405).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135492/

sarutak committed:
Resolve conflict.

SparkQA commented:
**[Test build #135517 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135517/testReport)** for PR 31601 at commit [`0af19a1`](https://github.com/apache/spark/commit/0af19a1227eb9f14d9b4ce6dffe1e720258381ed).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40097/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40097/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40097/

SparkQA commented:
**[Test build #135517 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135517/testReport)** for PR 31601 at commit [`0af19a1`](https://github.com/apache/spark/commit/0af19a1227eb9f14d9b4ce6dffe1e720258381ed).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `sealed trait PartitionSpec extends LeafExpression with Unevaluable `
  * `trait V2PartitionCommand extends Command `
  * `case class TruncateTable(table: LogicalPlan) extends Command `
  * `case class TruncatePartition(`
  * `case class TruncatePartitionExec(`

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135517/

sarutak commented:
@cloud-fan According to [this](https://github.com/apache/spark/pull/31569#issuecomment-786344739), should we wait for what happens as a result of the discussion?

cloud-fan commented:
Is there a released Scala version that forbids the symbol's literal syntax?

sarutak commented:
No, but we get warning with Scala 2.13.

cloud-fan commented:
OK seems the ship is already sailed. Let's wait a bit more but I'm not sure we can change that decision...

sarutak commented:
> OK seems the ship is already sailed. Let's wait a bit more but I'm not sure we can change that decision...

Yeah, I think so too. At least, symbol literal will be no longer supported.

sarutak committed:
Resolve conflict.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135691/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40273/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40273/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40273/

sarutak commented:
@cloud-fan Surprisingly, Scala 3 seems to implement the backward compatibility feature for symbol literals (it may be a life extension measure, however).
https://github.com/lampepfl/dotty/pull/11588

So, I think replacing symbol literals with `mapAttr` might not be needed.
How about changing `map` to `mapAttr`? Do you think it's not needed too?
If so, I'll just close this PR.

cloud-fan commented:
If scala will keep that backward compatibility feature forever, then we don't need to do anything. Otherwise, we still need to prepare for the removal of the symbol literal.

sarutak commented:
> f scala will keep that backward compatibility feature forever,

According to [this](https://github.com/lampepfl/dotty-feature-requests/issues/182#issuecomment-793103603), it seems to be a life extension measure and will be removed in the future.
But I'm asking the Scala community.
